File: src/python/hud.py
from browser import window
import ika
from thing import Thing

def sgn(i):
    if i > 0: return 1
    elif i < 0: return -1
    else: return 0

class _Gauge(Thing):
    def __init__(self, engineRef, imageName):
        '''
        imageName - name of the image series to use.
         ie 'gfx/ui/barhp%i.png'
        '''

        self.engineRef = engineRef
        (self.span, self.left, self.middle, self.right) =\
            [engineRef.getImage(imageName % i) for i in range(4)]
        self.oldVal = self.oldMax = 0
        self.x = 0
        self.y = 0
        self.opacity = 0
        self.rgb = [255, 255, 255]
        self.width = None
        self.fadeIn = False

    def update(self):
        v = sgn(self.curVal - self.oldVal)
        m = sgn(self.curMax - self.oldMax)

        if self.fadeIn:
            self.opacity = min(512, self.opacity + 20)
            if self.opacity == 512:
                self.fadeIn = False

        if not v and not m:
            self.opacity = max(0, self.opacity - 1)
        else:
            self.fadeIn = True
            self.oldVal += v
            self.oldMax += m

    def draw(self):
        if self.opacity == 0:
            return

        o = min(1.0, self.opacity / 255.0)

        # the width of the repeated span image thingo.
        # each end of the gauge occupies two pixels, so we subtract four.
        # (bad hack, I know)
        width = (self.width or self.oldMax) - 3

        x = self.engineRef.video.xres - width - self.left.width - self.right.width - self.x - 2

        self.engineRef.video.TintBlit(self.left, x, self.y, o)
        self.engineRef.video.TintBlit(self.right, x + width + self.left.width, self.y, o)

        x += self.left.width

        self.engineRef.video.TintScaleBlit(self.span, x, self.y, width, self.span.height, o)

        x -= 2

        if self.width:
            v = self.oldVal * self.width // self.oldMax
        else:
            v = self.oldVal

        if self.oldVal:
            self.drawRect(x + (self.width or self.oldMax) - v, self.y + 5, x + (self.width or self.oldMax), self.y + 6, o)

    def drawRect(self, x, y, w, h, opacity):
        'Used to draw in the filled part of the gauge.'
        self.engineRef.video.DrawRect(x, y, w, h, window.RGB(*(self.rgb + [opacity * 255])))

    curVal = property(lambda self: None) # ditto
    curMax = property(lambda self: None) # override.  Needs to be readable.

class HPBar(_Gauge):
    def __init__(self, engineRef):
        _Gauge.__init__(self, engineRef, 'gfx/ui/barhp%i.png')
        self.y = self.engineRef.video.yres - self.left.height - 1
        self.rgb = [255, 0, 0]

    curVal = property(lambda self: self.engineRef.player.stats.hp)
    curMax = property(lambda self: self.engineRef.player.stats.maxhp)

class MPBar(_Gauge):
    def __init__(self, engineRef):
        _Gauge.__init__(self, engineRef, 'gfx/ui/barhp%i.png')
        self.y = self.engineRef.video.yres - self.left.height * 2 - 1
        self.rgb = [0, 0, 255]
        self.oldMax = self.curMax
        self.oldVal = self.curVal

    curVal = property(lambda self: self.engineRef.player.stats.mp)
    curMax = property(lambda self: self.engineRef.player.stats.maxmp)

class EXPBar(_Gauge):
    def __init__(self, engineRef):
        _Gauge.__init__(self, engineRef, 'gfx/ui/barmp%i.png')
        #self.y = self.engineRef.video.yres - self.left.height * 2 - 1
        self.width = 100
        self.rgb = [0, 128, 128]
        self.oldMax = self.curMax
        self.oldVal = self.curVal

    def drawRect(self, x, y, w, h, opacity):
        super(EXPBar, self).drawRect(x, y, w, h - 1, opacity)

    curVal = property(lambda self: self.engineRef.player.stats.exp * self.width // self.engineRef.player.stats.next)
    curMax = property(lambda self: self.width)

File: src/winter/hud.ts
import { Thing } from "./thing.js"
import { Engine, Image, PyEngine } from "./winter.js"

const sgn = (i: number) => {
    if (i > 0) { return 1 }
    if (i < 0) { return -1 }
    return 0
}

class Gauge extends Thing {
    protected engine: Engine
    protected engineRef: PyEngine
    protected span: Image
    protected left: Image
    protected middle: Image
    protected right: Image
    protected oldVal: number
    protected oldMax: number
    protected x: number
    protected y: number
    protected opacity: number
    protected rgb: [number, number, number]
    protected fadeIn: boolean

    constructor(
        engineRef: PyEngine,
        imageNames: [string, string, string, string],
    ) {
        super()
        // imageName - name of the image series to use.
        //  ie 'gfx/ui/barhp%i.png'

        this.engine = engineRef.getEngine()
        this.engineRef = engineRef
        const images = imageNames.map(n => this.engine.getImage(n))
        ;[this.span, this.left, this.middle, this.right] = images
        this.oldVal = 0
        this.oldMax = 0
        this.x = 0
        this.y = 0
        this.opacity = 0
        this.rgb = [255, 255, 255]
        this.fadeIn = false
    }

    curVal() { return 0 }
    curMax() { return 0 }

    update() {
        const v = sgn(this.curVal() - this.oldVal)
        const m = sgn(this.curMax() - this.oldMax)

        if (this.fadeIn) {
            this.opacity = Math.min(512, this.opacity + 20)
            if (this.opacity === 512) {
                this.fadeIn = false
            }
        }

        if (v === 0 && m === 0) {
            this.opacity = Math.max(0, this.opacity - 1)
        } else {
            this.fadeIn = true
            this.oldVal += v
            this.oldMax += m
        }
        return false
    }

    draw() {
        if (this.opacity === 0) {
            return
        }

        const o = Math.min(1.0, this.opacity / 255.0)

        // the width of the repeated span image thingo.
        // each end of the gauge occupies two pixels, so we subtract four.
        // (bad hack, I know)
        const width = this.oldMax - 3

        let x = this.engine.video.xres - width - this.left.width - this.right.width - this.x - 2

        this.engine.video.tintBlit(this.left, x, this.y, o)
        this.engine.video.tintBlit(this.right, x + width + this.left.width, this.y, o)

        x += this.left.width

        this.engine.video.tintScaleBlit(this.span, x, this.y, width, this.span.height, o)

        x -= 2

        let v: number= this.oldVal

        if (this.oldVal) {
            this.drawRect(x + this.oldMax - v, this.y + 5, x + this.oldMax, this.y + 6, o)
        }
    }

    drawRect(x: number, y: number, w: number, h: number, opacity: number) {
        // Used to draw in the filled part of the gauge.
        this.engine.video.drawRect(x, y, w, h, this.rgb[0], this.rgb[1], this.rgb[2], opacity)
    }
}

export class HPBar extends Gauge {
    constructor(
        engineRef: PyEngine,
    ) {
        super(engineRef, [
            'gfx/ui/barhp0.png',
            'gfx/ui/barhp1.png',
            'gfx/ui/barhp2.png',
            'gfx/ui/barhp3.png',
        ])
        this.y = this.engine.video.yres - this.left.height - 1
        this.rgb = [1, 0, 0]
    }

    curVal() { return this.engineRef.getPlayerEntity().stats.hp }
    curMax() { return this.engineRef.getPlayerEntity().stats.maxhp }
}

export class MPBar extends Gauge {
    constructor(
        engineRef: PyEngine,
    ) {
        super(engineRef, [
            'gfx/ui/barhp0.png',
            'gfx/ui/barhp1.png',
            'gfx/ui/barhp2.png',
            'gfx/ui/barhp3.png',
        ])
        this.y = this.engine.video.yres - this.left.height * 2 - 1
        this.rgb = [0, 0, 1]
        this.oldMax = this.curMax()
        this.oldVal = this.curVal()
    }

    curVal() { return this.engineRef.getPlayerEntity().stats.mp }
    curMax() { return this.engineRef.getPlayerEntity().stats.maxmp }
}

export class EXPBar extends Gauge {
    constructor(
        engineRef: PyEngine,
    ) {
        super(engineRef, [
            'gfx/ui/barmp0.png',
            'gfx/ui/barmp1.png',
            'gfx/ui/barmp2.png',
            'gfx/ui/barmp3.png',
        ])
        //this.y = this.engine.video.yres - this.left.height * 2 - 1
        this.rgb = [0, 0.5, 0.5]
        this.oldMax = this.curMax()
        this.oldVal = this.curVal()
    }

    drawRect(x: number, y: number, w: number, h: number, opacity: number) {
        super.drawRect(x, y, w, h - 1, opacity)
    }

    curVal() { return Math.floor(this.engineRef.getPlayerEntity().stats.exp * 100 / this.engineRef.getPlayerEntity().stats.next) }
    curMax() { return 100 }
}
